### PR TITLE
Blogging Prompts: Fix Core Data concurrency issues in scheduler

### LINKED ISSUE #20842: Core Data: Multithreading violation when retrieving reminder information
Core Data: Multithreading violation when setting blogging reminders.

Could be the cause of crashes in the production: https://github.com/wordpress-mobile/WordPress-iOS/issues/20840

### Steps to reproduce the behavior

1. Set blogging reminders (as described in https://github.com/wordpress-mobile/WordPress-iOS/issues/20841)
2. Pass `-com.apple.CoreData.ConcurrencyDebug 1` arguments on launch
3. Open Jetpack app and log in
4. Observe Multithreading_Violation

It happens when `reschedulePromptRemindersIfNeeded` is triggered which checks for the notification settings and then accesses `BloggingPromptSettingsReminderDays.getActiveWeekdays` from a different thread

```
#0	0x00000001845323a0 in +[NSManagedObjectContext __Multithreading_Violation_AllThatIsLeftToUsIsHonor__] ()
#1	0x0000000184531f7c in _PFAssertSafeMultiThreadedAccess_impl ()
#2	0x0000000184530f6c in _sharedIMPL_pvfk_core_c ()
#3	0x0000000105f1b4f0 in BloggingPromptSettingsReminderDays.getActiveWeekdays() at /Users/povilasstaskus/Projects/WordPress-iOS/WordPress/Classes/Models/BloggingPromptSettingsReminderDays+CoreDataClass.swift:19
#4	0x0000000105556148 in closure #1 in BloggingPromptCoordinator.reschedulePromptRemindersIfNeeded(for:completion:) at /Users/povilasstaskus/Projects/WordPress-iOS/WordPress/Classes/ViewRelated/Blog/Blogging Prompts/BloggingPromptCoordinator.swift:129
#5	0x0000000105556328 in thunk for @escaping @callee_guaranteed (@guaranteed UNNotificationSettings) -> () ()

```

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/e7ccbd87-ca76-43ec-a3da-1624bd76a859

### PR DESCRIPTION
**Pull request by dvdchr**
Fixes #20842

> **Warning**:
> Please note that this targets the release/22.6 branch, which is already in code freeze.

## Description

When scheduling or rescheduling prompts, the related methods make a call to `UNNotificationCenter` to either request for push authorization or fetch notification settings. However, the notification center does not guarantee that the completion block will be executed in the caller's queue. Since the completion block code involves Core Data, this raises a multithreading violation when run in debug mode.

This PR fixes the issue by wrapping the Core Data-related code within these methods in the managed object context's queue.

Additionally, this also improves the thread-safety of `BloggingPromptsService`'s init method by ensuring that any access to Core Data properties is performed within that object's context queue.

## To test

- From Xcode, launch the Jetpack app with `-com.apple.CoreData.ConcurrencyDebug 1` launch argument.
- Log in with your WordPress.com account.
- Go to Site Menu > Site Settings.
- Set Blogging Reminders with prompts enabled.
- 🔎 Verify that the reminders are successfully set and that no multithreading violation is raised.
- Relaunch the Jetpack app from Xcode.
- 🔎 Verify that no multithreading violation is raised.
- Go to Site Menu > Site Settings.
- In the Blogging Reminders setting, uncheck all the days from the reminders.
- 🔎 Verify that the reminders are successfully set and that no multithreading violation is raised.

Additionally, please also smoke test Blogging Reminders — creating, updating, or removing schedules; with and without prompts.

## Regression Notes
1. Potential unintended areas of impact
Should be none. The logic is kept as close to the original as possible to minimize impact.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes, and the existing unit tests are passing.

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

N/A — this PR doesn't contain UI-related changes.

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/Services/BloggingPromptsService.swift [hunk 1]
```diff
@@ -179,17 +179,45 @@ class BloggingPromptsService {
...
+        guard let siteID,
+              let remoteInstance else {
             return nil
         }
 
         self.contextManager = contextManager
         self.siteID = siteID
-        self.remote = remote ?? .init(wordPressComRestApi: account.wordPressComRestV2Api)
+        self.remote = remoteInstance
     }
```
Comment (salimbraksa) [inline WordPress/Classes/Services/BloggingPromptsService.swift:220]: @dvdchr I'm wondering if it's possible to safely access both the `siteID` and the `WPAccount` instance using just the `mainContext`. cc @crazytonyli 

```swift
var remoteInstance = remote
var siteID: NSNumber? = nil

let blogObjectID = blog?.objectID
let context = contextManager.mainContext

context.performAndWait { 
     if let blogObjectID {
       let blogInContext = try? context.existingObject(with: blogObjectID)
       siteID = blogInContext?.dotComID
     }
     let account = try? WPAccount.lookupDefaultWordPressComAccount(in: context)  
     // ...
}

// Then we set `self` properties.
```

Your approach also looks correct to me. You access the `siteID` using the blog's context and the account using the `mainContext`, that's also fine I guess. I'll let you and other reviewers decide how to move forward.

File: WordPress/Classes/Services/BloggingPromptsService.swift [hunk 2]
```diff
@@ -179,17 +179,45 @@ class BloggingPromptsService {
...
+        }
+
+        // fetch the default account and fall back to default values as needed.
+        contextManager.performQuery { mainContext in
+            guard let account = try? WPAccount.lookupDefaultWordPressComAccount(in: mainContext) else {
+                return
+            }
+            siteID = siteID ?? account.primaryBlogID
+            remoteInstance = remoteInstance ?? .init(wordPressComRestApi: account.wordPressComRestV2Api)
+        }
```
Comment (crazytonyli) [inline WordPress/Classes/Services/BloggingPromptsService.swift:210]: `performQuery` can return a query result. You can potentially refactor the above code to something like this (pseudo code)

```swift
let blogID = blog.objectID
let (siteID, remoteInstance) = contextManager.performQuery { context in
  let blog: Blog? = context.existingObject(with: blogID)
  let account: Account? = WPAccount.lookupDefault()
  return (
    blog.dotComID ?? account.primaryBlogID,
    .init(wordPressComRestApi: account.wordPressComRestV2Api)
  )
}
```

File: WordPress/Classes/Utility/Blogging Prompts/PromptRemindersScheduler.swift [hunk 3]
```diff
@@ -43,9 +43,15 @@ class PromptRemindersScheduler {
     ///   - time: The user's preferred time to be notified.
     ///   - completion: Closure called after the process completes.
     func schedule(_ schedule: BloggingRemindersScheduler.Schedule, for blog: Blog, time: Date? = nil, completion: @escaping (Result<Void, Error>) -> Void) {
+        guard let context = blog.managedObjectContext else {
+            return
```
Comment (crazytonyli) [inline WordPress/Classes/Utility/Blogging Prompts/PromptRemindersScheduler.swift:48]: Do you think we should call the `completion` before returning? It's usually good to make sure completion is called once and once only.

**Review by crazytonyli — COMMENTED** [review]
Pattern like `blog.managedObjectContext.perform { doSomething(with: blog)}` feels like a patch to me. We probably should use it sparsely. Because there is no guarantee the `doSomething` function _only_ uses the passed-in `blog` instance in its calling thread.

For instance, a couple of call sites of `processSchedule` is changed to be called from the `blog` context thread. But the function itself accesses [the `blog` in an API callback which I believe is called from the main thread](https://github.com/wordpress-mobile/WordPress-iOS/blob/c5cf208ea8dde34b0666338efd31e12d9614d23a/WordPress/Classes/Utility/Blogging%20Prompts/PromptRemindersScheduler.swift#L188).

I think, yes, this change would fix the particular crash shown in https://github.com/wordpress-mobile/WordPress-iOS/issues/20848, but it may cause other issues.

One solution on top of my head is refactoring the Blog reminder classes to not use `Blog` type. It can declare a value type `struct BlogData` (with a better name probably), which includes a few property the Blog Reminder component really cares about: objectID, site id, titile, etc. So that it's not restricted by Core Data's programming model, but also has the capability to change the `Blog` model with the `objectID` if needed.

File: WordPress/Classes/Services/BloggingPromptsService.swift (hunk 1 above)
Comment (crazytonyli) [reply WordPress/Classes/Services/BloggingPromptsService.swift:220]: Agreed. We could use `coreDataStack.performQuery` function (which uses the main context) as I suggested in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854#discussion_r1228903792.

File: WordPress/Classes/Services/BloggingPromptsService.swift (hunk 2 above)
Comment (dvdchr) [reply WordPress/Classes/Services/BloggingPromptsService.swift:210]: Thanks. Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854/commits/efdad6e4a7db74ac3503246eae8cab9f8f14f332

File: WordPress/Classes/Utility/Blogging Prompts/PromptRemindersScheduler.swift (hunk 3 above)
Comment (dvdchr) [reply WordPress/Classes/Utility/Blogging Prompts/PromptRemindersScheduler.swift:48]: Good catch! This is addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854/commits/241e0ed5d0430a816bf76e76798cf2698b00ccff.

File: WordPress/Classes/Services/BloggingPromptsService.swift (hunk 1 above)
Comment (dvdchr) [reply WordPress/Classes/Services/BloggingPromptsService.swift:220]: Thanks for the input, folks 🙇🏼‍♂️. This should be addressed in https://github.com/wordpress-mobile/WordPress-iOS/commit/efdad6e4a7db74ac3503246eae8cab9f8f14f332.

**Comment by dvdchr** [comment]
Thanks for the input, @crazytonyli!

> [...] this change would fix the particular crash shown in https://github.com/wordpress-mobile/WordPress-iOS/issues/20848, but it may cause other issues.
> [...] One solution on top of my head is refactoring the Blog reminder classes to not use Blog type.

Great idea, and I agree. I was considering this yesterday but realized that there are going to be a lot of changes. Since I'm targeting 22.6 (which is in code freeze), I'm thinking of keeping the impact of this PR minimal and doing a larger refactor separately in the next iteration.

However, it doesn't make sense if the improvement is flawed; as you've pointed out:

> For instance, a couple of call sites of processSchedule is changed to be called from the blog context thread. But the function itself accesses [the blog in an API callback which I believe is called from the main thread](https://github.com/wordpress-mobile/WordPress-iOS/blob/c5cf208ea8dde34b0666338efd31e12d9614d23a/WordPress/Classes/Utility/Blogging%20Prompts/PromptRemindersScheduler.swift#L188).

Thanks for catching that. To address this, I've made some improvements to the `processSchedule` (along with a few other methods) in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854/commits/6f89aa7fd9c2d541f46cea7e486d36774bf236d7 so that the `Blog` object is no longer directly accessed from the API completion block. Although there are still some direct calls to the `Blog` object before the API call, at least the method is private and all of its callers ensure that the method is called from the `Blog`'s context queue.

In the (near) future, the larger refactor will remove the `blog` parameter from most of the Blogging Reminders and Blogging Prompts methods, replacing them with `siteID` or a value type as per your and Jeremy's suggestion.

Anyways, can you give this a second pass, please? Thank you!

File: WordPress/Classes/Services/BloggingPromptsService.swift [hunk 4]
```diff
@@ -179,17 +179,44 @@ class BloggingPromptsService {
...
+    ///   - remote: When supplied, the service will use the specified remote service.
+    ///     Otherwise, a remote service with the default account's credentials will be used.
+    ///   - blog: When supplied, the service will perform blogging prompts requests for this specified blog.
+    ///     Otherwise, this falls back to the default account's primary blog.
     required init?(contextManager: CoreDataStackSwift = ContextManager.shared,
                    remote: BloggingPromptsServiceRemote? = nil,
                    blog: Blog? = nil) {
-        guard let account = try? WPAccount.lookupDefaultWordPressComAccount(in: contextManager.mainContext),
-              let siteID = blog?.dotComID ?? account.primaryBlogID else {
+        let blogObjectID =  blog?.objectID
```
Comment (crazytonyli) [inline WordPress/Classes/Services/BloggingPromptsService.swift]: 🧐 There is an extra space here.

File: WordPress/Classes/Utility/Blogging Prompts/PromptRemindersScheduler.swift [hunk 5]
```diff
@@ -282,12 +291,11 @@ private extension PromptRemindersScheduler {
...
                                 with schedule: Schedule,
                                 time: Time,
-                                blog: Blog,
+                                siteID: Int,
                                 maxDays: Int = Constants.staticNotificationMaxDays) -> [String]? {
         guard case .weekdays(let weekdays) = schedule,
               maxDays > 0,
-              let maxDate = Calendar.current.date(byAdding: .day, value: maxDays, to: afterDate),
-              blog.dotComID != nil else {
+              let maxDate = Calendar.current.date(byAdding: .day, value: maxDays, to: afterDate) else {
```
Comment (crazytonyli) [inline WordPress/Classes/Utility/Blogging Prompts/PromptRemindersScheduler.swift:298]: Not related to this PR. But I didn't realise that [`Calendar.current`](https://developer.apple.com/documentation/foundation/calendar/2293438-current)—the calendar that user picked in the device's System Settings—is used a lot in the app. In most cases, we want the Gregorian calendar, not the calendar the user likes their phone to display.

**Review by crazytonyli — APPROVED** [review]
Thanks for addressing my feedback. This PR looks good to me. 👍

It would be nice to see a proper refactor done to the Blog Reminder part of the app soon though 😺

File: WordPress/Classes/Services/BloggingPromptsService.swift (hunk 4 above)
Comment (dvdchr) [reply WordPress/Classes/Services/BloggingPromptsService.swift]: Good eye! 🦅. Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20854/commits/a9d5782042c7cc1cc58a2a755c14d2cd1b084093.

File: WordPress/Classes/Utility/Blogging Prompts/PromptRemindersScheduler.swift (hunk 5 above)
Comment (dvdchr) [reply WordPress/Classes/Utility/Blogging Prompts/PromptRemindersScheduler.swift:298]: Hmm, yeah, I think I didn't really think about it and just followed how it's done across the codebase. I'll consider fixing this during the refactor later. Thanks!

**Comment by mokagio** [comment]
@crazytonyli I ended up here while looking into something else, just wanted to say that this sounds like a great idea:

> One solution on top of my head is refactoring the Blog reminder classes to not use Blog type. It can declare a value type struct BlogData (with a better name probably), which includes a few property the Blog Reminder component really cares about: objectID, site id, titile, etc. So that it's not restricted by Core Data's programming model, but also has the capability to change the Blog model with the objectID if needed.

**Pull request merged**